roles/ovirt-image-template/tasks/qcow2_image.yml

```diff
@@ -22,6 +22,27 @@
   tags:
     - ovirt-template-image
 
+- name: Check for qemu-img
+  stat: path=/usr/bin/qemu-img
```

Comment:
Hmm, this way we need to require `qemu-img` package. I think it's not shipped in channel on engine.
Maybe it would be better to create a template with disk which has size same as qcow disk image, and then resize it via oVirt API.

Then write a small script which find the size of qcow image, taking a look at[1], shouldn't be that hard.

[1] https://people.gnome.org/~markmc/qcow-image-format.html

Reply:
That is possible, I can read the size parameter directly with the following snippet:

    hexdump -s 24 -l 8 /tmp/ovirt_image_data

Is it the intent of the ovirt-ansible project to always be run (or able to be run) from the engine?

Reply:
In general it should be possible to run it on any machine. But we need to say what are the requirements. The current patch is OK, if we say that user have to run the role on the machine with installed `qemu-img`. But I think it's better to make it possible to run it even on clean el7 machine, if it's possible.

Reply:
I'm looking at other options, but one thing I'd like to add in is that `hexdump` is a bit clunky to script, and that the `xxd` tool from `vim-common` provides better output AND includes the ability to write to a binary file at an offset and length should we want to edit the file before uploading without `qemu-img`.

`vim-common` is also missing from an engine's standard install, (and the RHEL 7.4 KVM guest image) but is closer to a RHEL standard.

Reply:
I think writing a small python script would be the best approach, what do you think?

Reply:
How about this for a beginning?

```
#!/usr/bin/python

import sys
import struct

if __name__ == "__main__":
    try:
        target_gib = int(sys.argv[1])
        qcow_filename = sys.argv[2]
    except IndexError:
        print "Usage: " + sys.argv[0] + " <size in GiB> <filename>"
        sys.exit(1)
    with file(qcow_filename, 'r+b') as fh:
        fh.seek(24,0)
        current_packed = fh.read(8)
        (size_file,) = struct.unpack('>Q', current_packed)
        target_bytes = target_gib * 2**30
        if target_bytes > size_file:
            newsize_packed = struct.pack('>Q', target_bytes)
            fh.seek(24,0)
            fh.write(newsize_packed)
```

Reply:
I just wonder whether is it safe to rewrite the size of the qcow image like that.
I think it would be better to just get the size of the qcow image, and create a template with that image. Then check if `template_disk_size` is bigger and if yes call `ovirt_disks` module to handle resize.

Also I've just noticed that file[1] command is parsing the qcow files as well:

```bash
# file cirros-0.3.0-x86_64-disk.img 
cirros-0.3.0-x86_64-disk.img: QEMU QCOW Image (v2), 41126400 bytes

# file cirros-0.3.0-x86_64-disk.img | awk '{print $(NF-1)}'
41126400

```

[1] https://github.com/file/file/blame/a25f230f1e7adf1632592d5c81a2d6cc86559c67/magic/Magdir/virtual#L33

Reply:
calling the `ovirt_disks` should be just a matter of:

```yaml
ovirt_disks:
  id: "{{ ovirt_disk.id }}"
  auth: "{{ ovirt_auth }}"
  size: "{{ template_disk_size }}"
when: template_disk_size > qcow_disk_size_in_gb
```

Reply:
A diff of a `qemu-img` re-sized ovirt_image_data and one my python script re-sized comes up with no differences. All the complexity in that snippet of code seems to be around the truncate case, which we're purposely avoiding.

I can test with the api, but my initial problem was that re-sizing the template resulted in a correctly sized disk image on the RHV server (even all VMs are created with 60G root disks) but the OS does not see the extra size because the image is not getting adjusted, only the backing store.

Reply:
OK, I just wasn't sure if just re-writing that value is enough, if that's enough, let's use your script.
Thanks a lot!

Reply:
Additional testing has shown that my test image received more changes than just the size header when qemu-img re-sized it. If I run my script against a newly downloaded qcow2, it makes the file unreadable by qemu-img.

Reply:
I've also tested with the `ovirt_disks` method outlined earlier, and it does not affect the disk size. (The temporary VM disk remains at 10GiB despite attempts to re-size it to 60GiB.)

I also attempted to set the size parameter on template VM creation, also to no effect.